[KVConnector] Add KV events to KV Connectors

hickeyma opened the pull request:
## Purpose

Enable KV connectors to pass KV Events to vLLM. vLLM can then publish the events using its existing KV Events publisher in the engine. 

The motivation to emit KV Events by connectors is to help KV-cache-aware routing make more efficient decisions, based on not only KV caches in GPU, but also KV caches in CPU and local disk.

This PR includes updates to the the LMCache connector to implement the `KVConnectorKVEvents` interface that enables KV events to be passed to vLLM from LMCache. 

This is part of the solution for:
- RFC [[RFC]: KV-Cache Interoperability API Standardization ](https://github.com/vllm-project/vllm/issues/20492) which proposes "formalizing KV Events as the public contract for any component emitting or consuming KV-Cache lifecycle events, including external indexers, routers, and engines". 

Note: This is related to #28252 and from feedback by @markmc in https://github.com/vllm-project/vllm/pull/28252#issuecomment-3501398970

## Test Plan

`pytest tests/v1/kv_connector/unit/test_lmcache_connector.py`

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

hickeyma committed:
Send kv events from worker side to scheduler side

This is required for when worker side operations like CPU offloading
generate KV cache events. This commit enables theses events to be passed
to the scheduler side so that they can be published by the engine.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update after review from gemini-code

Update comments:
- https://github.com/vllm-project/vllm/pull/28309#pullrequestreview-3435326264

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on vllm/distributed/kv_transfer/kv_connector/v1/lmcache_connector.py:
```diff
@@ -183,6 +213,25 @@ def build_connector_meta(
+            or not kv_events.events
+        ):
+            return
+        self._kv_events = kv_events.events
```
Updated

hickeyma committed:
Remove change to the LMCache connector

The changes to the connector is for a separate PR and this
PR is independent of it for now.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

orozery commented:
@hickeyma I think you're missing aggregation code in `KVOutputAggregator`.
This means adding more state to `KVOutputAggregator` to aggregate per-block KVEvents.

@njhill Let's think how this change fits to the longer-term future.
Right now, scheduler can send arbitrary metadata to the workers (`KVConnectorMetadata`).
On the other hand, workers are limited to an explicit `KVConnectorOutput`, which includes:
- request-level completions (`finished_sending`, `finished_recving`)
- abstract stats
- block-level errors
- (just noticed this) `expected_finished_count` - looks like a niche bootstrap configuration param

At the time when I introduced worker->scheduler output aggregation (#19555), my initial suggestion was to allow abstract metadata to flow from worker connectors to the scheduler connector.
I think that the current approach where we limit the worker output has 2 drawbacks:
1. It makes it harder to implement connector logic (e.g. the current PR is needed).
2. For those connectors who does succeed in pushing their logic to `KVConnectorOutput`, the result is a growing-over-time `KVConnectorOutput` which contains a scattered list of fields which are practically connector-specific.

My view is that we should rename `KVConnectorStats` to `KVConnectorWorkerMeta` (which could support `get_stats() -> KVConnectorStats`).
This will allow free us from maintaining the list of connector-specific worker output fields, and delegate it entirely to the connector implementations themselves.

cc @sdavidbd

markmc commented:
> @njhill Let's think how this change fits to the longer-term future. Right now, scheduler can send arbitrary metadata to the workers (`KVConnectorMetadata`). On the other hand, workers are limited to an explicit `KVConnectorOutput`,

I agree with this in general - e.g. in an earlier iteration of #26172 I found myself adding NIXL-specific semantics to the contents of `finished_recving`, which was only possible because I happened to be able to use the `update_connector_output()` to extract the NIXL-specific data and translate the contents to what the scheduler expected. An obscure example, but I did think it was a weird contrast with the complete flexibility of connector-specific `KVConnectorMetadata` subclasses

> My view is that we should rename `KVConnectorStats` to `KVConnectorWorkerMeta` (which could support `get_stats() -> KVConnectorStats`). 

I'm not sure I agree with this, though:

1. Although `KVConnectorStats` is a recent addition, we should by default assume that we need to maintain compatibility and break any potential external connectors that might have already adopted
2. `KVConnectorStats` seems like a useful abstraction that should probably be adopted by all connectors over time
3. `KVConnectorStats` is very specifically about metrics and integration with Prometheus/console stat logging - for example the `aggregate()` and `reduce()` methods - so not easily repurposed as a generic worker->scheduler channel
4. `KVEventsBatch` also seems like a useful, connector-independent abstraction, that ties in nicely with the existing public KV events publishing API, so if connectors have to use a connector-specific metadata channel for them, we'd probably see a lot of duplication

But I do think it could be positive to add `KVConnectorWorkerMetdata` anyway ...

hickeyma commented:
Thanks @njhill @orozery @markmc @NickLucche for the feedback. I will get to it in due course as bit delayed this week because I'm at KubeCon.

> Honestly, I don't yet fully understand the motivation for the lmcache connector to emit its own events (as per https://github.com/vllm-project/vllm/pull/28252 and https://github.com/LMCache/LMCache/issues/1846) in addition to the KV events already emitted by vLLM ... so I don't have a strong sense either way whether this need is highly-specific to lmcache

@markmc The reason that a connector like LMCache would need to generate their own events is down to specific details or information. How LMcache stores the tokens/hashes (e.g. LMcache stores tokens with a granularity of 256 as a block and vLLM granularity is 16 as a block) and where it stores (e.g. LMcache might store them to disk) them will differ to vLLM. Therefore, a consumer of the events would be missing the full context of the cache if it only consumed the vLLM events.

hickeyma commented:
> I understand your point here, but I believe that adding more fields to KVConnectorOutput was kinda part of the original design where it was expected that it would naturally grow to contain more stuff (I think this PR is a good example of that).
Naturally nobody would want the struct to grow indefinitely if we sense that this is the direction in which we're heading.

I agree with @NickLucche about having a separate field. @orozery I understand you reservation about only 1 connector possibly using it but still think the separation is best to clearly identify it.

ApostaC commented on vllm/distributed/kv_transfer/kv_connector/v1/base.py:
```diff
@@ -343,6 +343,12 @@ def get_kv_connector_stats(self) -> Optional["KVConnectorStats"]:
+    def get_kv_connector_kv_cache_events(self) -> Optional["KVEventBatch"]:
+        """
+        Get the KV connector kv cache events collected during the last interval.
+        """
```
Can we be more specific about whether the scheduler or the worker calls this function?

hickeyma replied:
@ApostaC Its part of the "Worker-side methods" section of https://github.com/vllm-project/vllm/blob/main/vllm/distributed/kv_transfer/kv_connector/v1/base.py. Are you asking for more context in the method doc string to include that it is called by the mode runner and when?

Maybe a sentence like the following:

`This function should be called by the model runner every time after the model execution and before cleanup.`?

hickeyma replied:
Clarification added to method doc string.

hickeyma committed:
Remove changes to LMCache connector

LMCache connector changes are not part of this PR as this PR deals
with adding ability to pass KV events between worker and scheduler sides
and not a specific connector usage of this.

The connector implementation is in a separate PR.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Combine events from different worker connectors

It is part of the aggregation of kv_connector_output from
all workers. For KV cache events, this means combining events
from all workers, remvoing any duplications.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add comment to get_kv_connector_kv_cache_events for clarity

Clarify that it is the workser side that calls the method.

Review comment:

- https://github.com/vllm-project/vllm/pull/28309#discussion_r2519956129

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

ApostaC reviewed (commented):
Otherwise LGTM

Hey @orozery please also take a look

orozery commented:
I'm trying to understand `combine_unique_ordered_events`:
Each worker can output, e.g., `BlocksStored`.
Assuming a single worker stored blocks for a specific block. Wouldn't you need to somehow count that all workers also stored the same block before reporting this event?

hickeyma committed:
Remove equality methods in KV Cache events data classes

Equality methods are unnecessary as its provided by the parent class
'msgspec.Struct' which they extend from.

Review comment:
https://github.com/vllm-project/vllm/pull/28309/files#r2539044714

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sdavidbd commented:
> > Also note that the current PR does not add any Scheduler logic to process the kv_cache_events field. As a result, this data is never consumed and the field is effectively unused. It seems that some Scheduler handling would be needed before this can be considered ready for merge.
> 
> @sdavidbd See #28252

I stand corrected.

hickeyma committed:
Add improvements from code review comments

Review comments:
- https://github.com/vllm-project/vllm/pull/28309#discussion_r2545498944
- https://github.com/vllm-project/vllm/pull/28309#discussion_r2545587736

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Handle duplicate kv events from workers

Workers will generate duplicate kv events in LMCache. This
commit adds capability to aggregate on events from workers by returning
only those that were emitted by all workers.

It also provide an abstract class KVConnectorKVEevnts that is implemented
by connectors to handle how they emit events.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

KuntaiDu reviewed (approved):
LGTM! @NickLucche could you please also take a look?

orozery commented on vllm/v1/worker/kv_connector_model_runner_mixin.py:
```diff
@@ -138,17 +137,17 @@ def _get_kv_connector_output(
-            output.kv_connector_stats = (
-                KVConnectorModelRunnerMixin.get_kv_connector_stats()
-            )
+            if has_kv_transfer_group():
```
I don't think this check is necessary

orozery commented on vllm/v1/worker/kv_connector_model_runner_mixin.py:
```diff
@@ -138,17 +137,17 @@ def _get_kv_connector_output(
-            )
+            if has_kv_transfer_group():
+                output.kv_connector_stats = (
+                    get_kv_transfer_group().get_kv_connector_stats()
```
I think you can replace `get_kv_transfer_group()` with `kv_connector` here and down below.

hickeyma replied:
Good suggestion. Done.

orozery reviewed (commented):
@hickeyma Thanks for your hard work!

I think that the changes introduced here add up code to vLLM that could have reside inside the LMCache connector implementation.
Specifically, the changes to `KVConnectorOutput`, `KVConnectorBase_V1` and the introduction of `KVEventAggregator`.
I do not think this changes, although trying to be generic, will fit any other use-case.

From another viewpoint, I expect `KVConnectorOutput` to contain only fields that are of interest to vLLM's scheduler (e.g. `finished_sending`, `finished_recving`, `invalid_block_ids`, etc.). 
The new `KVConnectorOutput.kv_cache_events` field introduced here is only consumed by the LMCache connector. From the vLLM point of view it's a don't care.
I think that making this field opaque behind a connector-specific metadata is the right solution.

I discussed this in person with @hickeyma a few weeks ago.
I think the bottom-line was:
1. @hickeyma supports a solution along the lines I've drawn.
2. @hickeyma prefers that we proceed with the current approach as a start, and perhaps change the implementation later.
3. I said I have no strong objection with merging the current approach.

I don't think taking the other approach requires a lot of coding effort.
I would still prefer making these changes now, instead of introducing API changes that may revert in the future.

Anyhow, please don't consider my opinion as a blocker for proceeding here.

hickeyma committed:
Updates after review

Review comments:
- https://github.com/vllm-project/vllm/pull/28309#discussion_r2580395340
- https://github.com/vllm-project/vllm/pull/28309#discussion_r2580397162
- https://github.com/vllm-project/vllm/pull/28309#discussion_r2580428717

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thank you @orozery for review, comments and feedback. I have pushed changes for your review comments.

> I think that the changes introduced here add up code to vLLM that could have reside inside the LMCache connector implementation.
Specifically, the changes to KVConnectorOutput, KVConnectorBase_V1 and the introduction of KVEventAggregator.
I do not think this changes, although trying to be generic, will fit any other use-case.
From another viewpoint, I expect KVConnectorOutput to contain only fields that are of interest to vLLM's scheduler (e.g. finished_sending, finished_recving, invalid_block_ids, etc.).
The new KVConnectorOutput.kv_cache_events field introduced here is only consumed by the LMCache connector. From the vLLM point of view it's a don't care.
I think that making this field opaque behind a connector-specific metadata is the right solution.

I appreciate your feedback and the points raised with regard to this. I wanted however to follow the pattern currently in place for stats. For stats, changes were added to `KVConnectorOutput`, `KVConnectorBase_V1`, and `KVConnectorStats` was introduced. I think it is a flexible design to make this generic for now which gives other connectors the ability to provide KV events if they want to.

orozery commented on vllm/distributed/kv_transfer/kv_connector/utils.py:
```diff
@@ -201,16 +202,36 @@ def update_finished_set(
+        # events for all workers.
+        if combined_kv_cache_events is not None:
+            combined_kv_cache_events.set_workers(self._expected_finished_count)
+            combined_kv_cache_events = combined_kv_cache_events.aggregate()
```
`combined_kv_cache_events` holds up information on worker events on this step.
Events that did not accumulate for enough workers will be lost after this step.
You need to keep track of those events cross-steps.
See for example how they do it for `finished_recving` (using `self._recv_remaining_count`).

hickeyma replied:
That's interesting, thanks for letting me know and explaining this on slack.

I have pushed 717702fe5d6a41c6607ffe8c5c2736674e47ca86 which persists events across steps which is requied if workers are not processed in the same step. It then resets the events once the scheduler retrieves them in `take_events()`.

hickeyma committed:
Handle events for workers not processed at same time

All workers may not be processed by the same instance running
`aggregate()` for the model runner. Therefore, events needs to be persisted
across KVConnector output processing until the events have been retrieved
in `take_events()`. This commit persisted the events until all worker have been
completed.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

NickLucche reviewed (commented):
Thanks for the work @hickeyma , and apologies for the delay in getting back to this PR.
This is an LGTM re: interface changes for me.

I don't mean to delay this contrib further, however it would be really nice to have any sort of unit tests here, especially for the MultiConnector case.

hickeyma committed:
Add unit tests

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

ApostaC commented:
Hey @hickeyma , can you do "update branch" to include the fix of LMCache unit tests? Otherwise, this looks good to me!

ApostaC reviewed (approved):
Thanks for the contribution and thanks @KuntaiDu @NickLucche for reviewing!

hickeyma committed:
Add lmcache as depoendency for unit test

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

NickLucche commented on requirements/test.txt:
```diff
@@ -506,6 +525,10 @@ networkx==3.2.1
     #   scikit-image
     #   torch
+nixl==0.8.0
+    # via lmcache
```
this version is going to be conflicting with the one we have in requirements/kv_connectors.txt
Any chance we can avoid installing nixl through lmcache?

NickLucche commented on requirements/test.txt:
```diff
@@ -1017,8 +1055,12 @@ sentence-transformers==3.2.1
+    #   setuptools-scm
     #   torch
+setuptools-scm==9.2.2
+    # via lmcache
```
same thing for scm here tbh

NickLucche commented on tests/v1/kv_connector/unit/test_lmcache_connector.py:
```diff
@@ -0,0 +1,756 @@
+from vllm.distributed.kv_transfer.kv_connector.v1.lmcache_connector import (
+    LMCacheConnectorV1,
+    LMCacheKVEvents,
+)
```
qq: is the lmcache package actually needed to run these tests or is it just for importing the types to insert in the mocks?

hickeyma replied:
Good catch, thanks @NickLucche. Updated to do a lazy import in the connector and therefore not requiring lmcache package during the mocking.

NickLucche reviewed (commented):
I think the only missing part here is to provide a generic implementation of `get_kv_connector_kv_cache_events` for the MultiConnector, such that it handles getting these events from multiple connectors, handling the case where only a subset of the requested connectors implement the `get_kv_connector_kv_cache_events` method (+quick test case).

I believe changes to `multi_connector.py` here https://github.com/vllm-project/vllm/pull/22188/ could be a good reference. 
@hickeyma

hickeyma committed:
Remove lmcache dependency in unit test

Review comment:
https://github.com/vllm-project/vllm/pull/28309#discussion_r2609722695

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add todo to multi_connector for adding KV events later

Comment:
https://github.com/vllm-project/vllm/pull/28309#pullrequestreview-3566351082

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Merge branch 'main' into send-events-worker-to-scheduler

NickLucche reviewed (approved):
LGTM as per @hickeyma comment on follow up work
